File: ch1/clip1/boj11279/Main.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        MaxPriorityQueue pq = new MaxPriorityQueue();
        int n = sc.nextInt();
        StringBuilder ans = new StringBuilder();
        while(n-- > 0) {
            int x = sc.nextInt();
            if(x == 0) {
                ans.append(pq.pop()).append('\n');
            } else {
                pq.push(x);
            }
        }
        System.out.println(ans);
    }
}

class MaxPriorityQueue {
    private int[] heap;
    private int size;

    public MaxPriorityQueue() {
        heap = new int[100001];
        size = 0;
    }

    public void push(int x) {
        heap[++size] = x;
        int current = size;
        while(current > 1) {
            int parent = current / 2;
            if(heap[parent] >= heap[current]) break;
            swap(parent, current);
            current = parent;
        }
    }

    public int pop() {
        if(size == 0) return 0;
        int ret = heap[1];
        heap[1] = heap[size--];
        int current = 1;
        while(current * 2 <= size) {
            int left = current * 2;
            int right = current * 2 + 1;
            int child = left;
            if(right <= size && heap[left] < heap[right]) {
                child = right;
            }
            if(heap[current] >= heap[child]) break;
            swap(current, child);
            current = child;
        }
        return ret;
    }

    private void swap(int a, int b) {
        int tmp = heap[a];
        heap[a] = heap[b];
        heap[b] = tmp;
    }
}

File: ch1/clip1/boj11279/main.cpp
#include <iostream>

int heap[100001];
int heapSize = 0;

void push(int num)
{
    heapSize++;
    heap[heapSize] = num;

    for(int index = heapSize; index > 1; index /= 2)
    {
        if(heap[index / 2] < heap[index])
        {
            std::swap(heap[index / 2], heap[index]);
        }
        else
        {
            break;
        }
    }
    return;
}

void pop()
{
    heap[1] = heap[heapSize];
    heapSize--;

    for(int index = 1; index <= heapSize;)
    {
        if(index * 2 <= heapSize)
        {
            int nextIndex = index * 2;

            if(index * 2 + 1 <= heapSize && heap[index * 2] < heap[index * 2 + 1])
            {
                nextIndex = index * 2 + 1;
            }
            
            if(heap[index] < heap[nextIndex])
            {
                std::swap(heap[index], heap[nextIndex]);
                index = nextIndex;
            }
            else
            {
                break;
            }
        }
        else
        {
            break;
        }
    }

    return;
}

int main()
{
    int N;
    scanf("%d", &N);

    for(int i = 0; i < N; i++)
    {
        int num;
        scanf("%d", &num);

        if(0 == num)
        {
            if(0 == heapSize)
            {
                printf("0\n");
            }
            else
            {
                printf("%d\n", heap[1]);
                pop();
            }
        }
        else
        {
            push(num);
        }
    }

    return 0;
}